jerry-main/main-unix-snapshot.c

```diff
@@ -0,0 +1,308 @@
+static uint32_t output_buffer[ JERRY_BUFFER_SIZE / 4 ];
+
+/**
+ * Loading a single file into the memory.
+ *
+ * @return true - if loading is successful
+ *         false - otherwise
+ */
+static size_t
+read_file (uint8_t *input_pos_p, /**< next position in the input buffer */
+           const char *file_name) /**< file name */
+{
+  FILE *file = fopen (file_name, "r");
+
+  if (file == NULL)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to open file: %s\n", file_name);
+    return 0;
+  }
+
+  size_t max_size = (size_t) (input_buffer + JERRY_BUFFER_SIZE - input_pos_p);
+
+  size_t bytes_read = fread (input_pos_p, 1u, max_size, file);
+  fclose (file);
+
+  if (bytes_read == 0)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to read file: %s\n", file_name);
+    return 0;
+  }
+
+  if (bytes_read >= max_size)
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: file too large: %s\n", file_name);
+    return 0;
+  }
+
+  printf ("Input file '%s' (%d bytes) loaded.\n", file_name, (int) bytes_read);
+  return bytes_read;
+} /* read_file */
+
+/**
+ * Merge command line option IDs
+ */
+typedef enum
+{
+  OPT_MERGE_HELP,
+  OPT_MERGE_OUT,
+} merge_opt_id_t;
+
+/**
+ * Merge command line options
+ */
+static const cli_opt_t merge_opts[] =
+{
+  CLI_OPT_DEF (.id = OPT_MERGE_HELP, .opt = "h", .longopt = "help",
+               .help = "print this help and exit"),
+  CLI_OPT_DEF (.id = OPT_MERGE_OUT, .opt = "o",
+               .help = "specify output file name (default: merged.snapshot)"),
+  CLI_OPT_DEF (.id = CLI_OPT_DEFAULT, .meta = "FILE",
+               .help = "input snapshot files, minimum two")
+};
+
+/**
+ * Process 'merge' command.
+ *
+ * @return error code (0 - no error)
+ */
+static int
+process_merge (cli_state_t *cli_state_p, /**< cli state */
+               int argc, /**< number of arguments */
+               char *prog_name_p) /**< program name */
+{
+  jerry_init (JERRY_INIT_EMPTY);
+
+  uint8_t *input_pos_p = input_buffer;
+
+  cli_change_opts (cli_state_p, merge_opts);
+
+  const char *output_file_name_p = "merged.snapshot";
+  const uint32_t *merge_buffers[argc];
+  size_t merge_buffer_sizes[argc];
+  uint32_t number_of_files = 0;
+
+  for (int id = cli_consume_option (cli_state_p); id != CLI_OPT_END; id = cli_consume_option (cli_state_p))
+  {
+    switch (id)
+    {
+      case OPT_MERGE_HELP:
+      {
+        cli_help (prog_name_p, "merge", merge_opts);
+        return JERRY_STANDALONE_EXIT_CODE_OK;
+      }
+      case OPT_MERGE_OUT:
+      {
+        output_file_name_p = cli_consume_string (cli_state_p);
+        break;
+      }
+      case CLI_OPT_DEFAULT:
+      {
+        const char *file_name_p = cli_consume_string (cli_state_p);
+
+        if (cli_state_p->error == NULL)
+        {
+          size_t size = read_file (input_pos_p, file_name_p);
+
+          if (size == 0)
+          {
+            return JERRY_STANDALONE_EXIT_CODE_FAIL;
```

Comment:
some error diagnostics would be nice here before bailing out from the program

Reply:
The read_file prints the error. We just acknowledge the fact.